apps/server/src/modules/h5p-editor/service/temporary-file-storage.service.ts

```diff
@@ -0,0 +1,128 @@
+		this.checkFilename(filename);
+		return this.repo.findByUserAndFilename(userId, filename);
+	}
+
+	public async deleteFile(filename: string, userId: string): Promise<void> {
+		this.checkFilename(filename);
+		const meta = await this.repo.findByUserAndFilename(userId, filename);
+		await this.s3Client.delete([this.getFilePath(userId, filename)]);
+		await this.repo.delete(meta);
+	}
+
+	public async fileExists(filename: string, user: IUser): Promise<boolean> {
+		this.checkFilename(filename);
+		const files = await this.repo.findAllByUserAndFilename(user.id, filename);
+		const exists = files.length !== 0;
+		return exists;
+	}
+
+	public async getFileStats(filename: string, user: IUser): Promise<H5pEditorTempFile> {
+		return this.getFileInfo(filename, user.id);
+	}
+
+	public async getFileStream(
+		filename: string,
+		user: IUser,
+		rangeStart = 0,
+		rangeEnd?: number | undefined
+	): Promise<Readable> {
+		this.checkFilename(filename);
+		const tempFile = await this.repo.findByUserAndFilename(user.id, filename);
+		const path = this.getFilePath(user.id, filename);
+		let rangeEndNew = 0;
+		if (rangeEnd === undefined) {
+			rangeEndNew = tempFile.size - 1;
+		}
+		const response = await this.s3Client.get(path, `${rangeStart}-${rangeEndNew}`);
+
+		return response.data;
+	}
+
+	public async listFiles(user?: IUser): Promise<ITemporaryFile[]> {
+		// method is expected to support listing all files in database
+		// Lumi uses the variant without a user to search for expired files, so we only return those
+
+		let files: ITemporaryFile[];
+		if (user) {
+			files = await this.repo.findByUser(user.id);
+		} else {
+			files = await this.repo.findExpired();
+		}
+
+		return files;
+	}
+
+	public async saveFile(
+		filename: string,
+		dataStream: ReadStream,
+		user: IUser,
+		expirationTime: Date
+	): Promise<ITemporaryFile> {
+		this.checkFilename(filename);
+		const now = new Date();
+		if (expirationTime < now) {
+			throw new NotAcceptableException('expirationTime must be in the future');
+		}
+
+		const path = this.getFilePath(user.id, filename);
+		let tempFile: H5pEditorTempFile | undefined;
+		try {
+			tempFile = await this.repo.findByUserAndFilename(user.id, filename);
+			await this.s3Client.delete([path]);
+		} finally {
+			if (tempFile === undefined) {
+				tempFile = new H5pEditorTempFile({
+					filename,
+					ownedByUserId: user.id,
+					expiresAt: expirationTime,
+					birthtime: new Date(),
+					size: dataStream.bytesRead,
+				});
+			} else {
+				tempFile.expiresAt = expirationTime;
+				tempFile.size = dataStream.bytesRead;
+			}
+		}
+		await this.s3Client.create(
+			path,
+			new H5pFileDto({ name: path, mimeType: 'application/octet-stream', data: dataStream })
+		);
+		await this.repo.save(tempFile);
+
+		return tempFile;
+	}
+
+	private getFilePath(userId: string, filename: string): string {
+		if (!userId || !filename) {
+			throw new ForbiddenException('COULD_NOT_CREATE_PATH');
```

Comment:
is forbidden the correct code here? is a request without userId of filename valid? is it even possible to call the function without a userId or filename?

If userId and filename is supplied by the user I would expect validation to fail, and a BadRequest to be thrown (in the controller layer). If this is coming from internal logic, something is very very broken, and this would appear to be an InternalServerError.

Anyway if its possible to pass something else than strings into the function, the types should be adjusted accordingly, eg. string | undefined.

Reply:
it's absolutly unnessary condithion cause it's not possible.